### ISSUE TITLE
Support querying repo mappings

### ISSUE THREAD

**Issue by cameron-martin** [bazelbuild/bazel#20631]
### Description of the feature request:

Add a way of querying for repo mappings, for usage in external tools (such as language servers). The repo mapping is necessary for these tools to resolve labels to files on disk, since the files live under the canonical repo name.

### Which category does this issue belong to?

External Dependency

### What underlying problem are you trying to solve with this feature?

Facebook have a [starlark parser & lsp](https://github.com/facebookexperimental/starlark-rust) which works pretty great with Bazel. However, it doesn't support bzlmod because it assumes the repo name of a label is the canonical name, i.e. it looks up `@foo` in `$(bazel info output_base)/external/foo`.

According to @fmeum, this is needed for intellij too.

### Any other information, logs, or outputs that you want to share?

Slack discussion: https://bazelbuild.slack.com/archives/C014RARENH0/p1703107424515999

**Comment by fmeum** [comment]
I started working on this in the past in https://github.com/bazelbuild/bazel/pull/18990, but noticed a problems with my implementation: All repositories generated by a module extension share the same repository mapping containing all these repos. If stored naively, this results in memory usage that is quadratic in the number of extension repos, which can easily be in the 1,000s. Thus, I think it's necessary to "intern" the repo mappings by having one array containing the deduplicated mappings and a map that references them by ID, e.g.:
```json
{
  "mappings": [{"foo": "foo~1.2.3, ...}, {"bar": "bar~4.5.6", ...}, ...],
  "repoToMappingIndex": {"foo~1.2.3": 0, ...}
}
```

**Comment by fmeum** [comment]
CC @mai93 as this format will also be needed to make IntelliJ's "Go to Definition" features work for labels with Bzlmod.

**Comment by meteorcloudy** [comment]
Looks like this one is done! \o/

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
